File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/settings/ToolSettingsViewModel.kt
package com.runanywhere.runanywhereai.presentation.settings

import android.app.Application
import android.content.Context
import timber.log.Timber
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.runanywhere.sdk.public.extensions.LLM.ToolDefinition
import com.runanywhere.sdk.public.extensions.LLM.ToolParameter
import com.runanywhere.sdk.public.extensions.LLM.ToolParameterType
import com.runanywhere.sdk.public.extensions.LLM.ToolValue
import com.runanywhere.sdk.public.extensions.LLM.ToolCallFormat
import com.runanywhere.sdk.public.extensions.LLM.RunAnywhereToolCalling
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Tool Settings UI State
 */
data class ToolSettingsUiState(
    val toolCallingEnabled: Boolean = false,
    val registeredTools: List<ToolDefinition> = emptyList(),
    val isLoading: Boolean = false,
)

/**
 * Tool Settings ViewModel
 *
 * Manages tool calling configuration and demo tool registration.
 * Mirrors iOS ToolSettingsViewModel.swift functionality.
 */
class ToolSettingsViewModel private constructor(application: Application) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(ToolSettingsUiState())
    val uiState: StateFlow<ToolSettingsUiState> = _uiState.asStateFlow()

    private val prefs = application.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    val toolCallingEnabled: Boolean
        get() = _uiState.value.toolCallingEnabled

    companion object {
        private const val PREFS_NAME = "tool_settings"
        private const val KEY_TOOL_CALLING_ENABLED = "tool_calling_enabled"
        
        /** Timeout for weather API requests (covers geocoding + weather fetch) */
        private const val WEATHER_API_TIMEOUT_MS = 15_000L

        @Volatile
        private var instance: ToolSettingsViewModel? = null

        fun getInstance(application: Application): ToolSettingsViewModel {
            return instance ?: synchronized(this) {
                instance ?: ToolSettingsViewModel(application).also { instance = it }
            }
        }
    }

    init {
        // Load saved preference
        val enabled = prefs.getBoolean(KEY_TOOL_CALLING_ENABLED, false)
        _uiState.update { it.copy(toolCallingEnabled = enabled) }

        // Refresh registered tools
        viewModelScope.launch {
            refreshRegisteredTools()
        }
    }

    fun setToolCallingEnabled(enabled: Boolean) {
        prefs.edit().putBoolean(KEY_TOOL_CALLING_ENABLED, enabled).apply()
        _uiState.update { it.copy(toolCallingEnabled = enabled) }
    }

    suspend fun refreshRegisteredTools() {
        val tools = RunAnywhereToolCalling.getRegisteredTools()
        _uiState.update { it.copy(registeredTools = tools) }
    }

    /**
     * Register demo tools matching iOS implementation:
     * - get_weather: Uses Open-Meteo API (free, no API key)
     * - get_current_time: Returns system time with timezone
     * - calculate: Evaluates math expressions
     */
    fun registerDemoTools() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            try {
                // Weather Tool - Uses Open-Meteo API (free, no API key required)
                RunAnywhereToolCalling.registerTool(
                    definition = ToolDefinition(
                        name = "get_weather",
                        description = "Gets the current weather for a given location using Open-Meteo API",
                        parameters = listOf(
                            ToolParameter(
                                name = "location",
                                type = ToolParameterType.STRING,
                                description = "City name (e.g., 'San Francisco', 'London', 'Tokyo')",
                                required = true
                            )
                        ),
                        category = "Utility"
                    ),
                    executor = { args: Map<String, ToolValue> ->
                        fetchWeather((args["location"] as? ToolValue.StringValue)?.value ?: "San Francisco")
                    }
                )

                // Time Tool - Real system time with timezone
                RunAnywhereToolCalling.registerTool(
                    definition = ToolDefinition(
                        name = "get_current_time",
                        description = "Gets the current date, time, and timezone information",
                        parameters = emptyList(),
                        category = "Utility"
                    ),
                    executor = { _: Map<String, ToolValue> ->
                        val now = Date()
                        val dateFormatter = SimpleDateFormat("EEEE, MMMM d, yyyy 'at' h:mm:ss a", Locale.getDefault())
                        val timeFormatter = SimpleDateFormat("HH:mm:ss", Locale.getDefault())
                        val isoFormatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.getDefault()).apply {
                            timeZone = TimeZone.getTimeZone("UTC")
                        }
                        val tz = TimeZone.getDefault()

                        mapOf(
                            "datetime" to ToolValue.StringValue(dateFormatter.format(now)),
                            "time" to ToolValue.StringValue(timeFormatter.format(now)),
                            "timestamp" to ToolValue.StringValue(isoFormatter.format(now)),
                            "timezone" to ToolValue.StringValue(tz.id),
                            "utc_offset" to ToolValue.StringValue(tz.getDisplayName(false, TimeZone.SHORT))
                        )
                    }
                )

                // Calculator Tool - Math evaluation
                RunAnywhereToolCalling.registerTool(
                    definition = ToolDefinition(
                        name = "calculate",
                        description = "Performs math calculations. Supports +, -, *, /, and parentheses",
                        parameters = listOf(
                            ToolParameter(
                                name = "expression",
                                type = ToolParameterType.STRING,
                                description = "Math expression (e.g., '2 + 2 * 3', '(10 + 5) / 3')",
                                required = true
                            )
                        ),
                        category = "Utility"
                    ),
                    executor = { args: Map<String, ToolValue> ->
                        val expression = (args["expression"] as? ToolValue.StringValue)?.value
                            ?: (args["input"] as? ToolValue.StringValue)?.value
                            ?: "0"
                        evaluateMathExpression(expression)
                    }
                )

                Timber.i("✅ Demo tools registered")
                refreshRegisteredTools()

            } catch (e: Exception) {
                Timber.e(e, "Failed to register demo tools")
            } finally {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    fun clearAllTools() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                RunAnywhereToolCalling.clearTools()
                refreshRegisteredTools()
                Timber.i("✅ All tools cleared")
            } catch (e: Exception) {
                Timber.e(e, "Failed to clear tools")
            } finally {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    /**
     * Detect the appropriate tool call format based on model name.
     * LFM2-Tool models use the lfm2 format, others use default JSON format.
     */
    fun detectToolCallFormat(modelName: String?): ToolCallFormat {
        val name = modelName?.lowercase() ?: return ToolCallFormat.Default
        return if (name.contains("lfm2") && name.contains("tool")) {
            ToolCallFormat.LFM2
        } else {
            ToolCallFormat.Default
        }
    }

    // Tool Executor Implementations

    /**
     * Fetch weather using Open-Meteo API (free, no API key required)
     * 
     * Uses a 15-second timeout for the entire operation (geocoding + weather fetch)
     * to ensure tool execution respects LLM timeout settings.
     */
    private suspend fun fetchWeather(location: String): Map<String, ToolValue> {
        return withContext(Dispatchers.IO) {
            try {
                // 15 second timeout for entire weather fetch operation
                // This covers both geocoding and weather API calls
                withTimeout(WEATHER_API_TIMEOUT_MS) {
                    // First, geocode the location
                    val geocodeUrl = "https://geocoding-api.open-meteo.com/v1/search?name=${URLEncoder.encode(location, "UTF-8")}&count=1"
                    val geocodeResponse = fetchUrl(geocodeUrl)

                    // Parse geocode response (simple JSON parsing)
                    val latMatch = Regex("\"latitude\":\\s*(-?\\d+\\.?\\d*)").find(geocodeResponse)
                    val lonMatch = Regex("\"longitude\":\\s*(-?\\d+\\.?\\d*)").find(geocodeResponse)
                    val nameMatch = Regex("\"name\":\\s*\"([^\"]+)\"").find(geocodeResponse)

                    if (latMatch == null || lonMatch == null) {
                        return@withTimeout mapOf(
                            "error" to ToolValue.StringValue("Location not found: $location"),
                            "location" to ToolValue.StringValue(location)
                        )
                    }

                    val lat = latMatch.groupValues[1]
                    val lon = lonMatch.groupValues[1]
                    val resolvedName = nameMatch?.groupValues?.get(1) ?: location

                    // Fetch weather
                    val weatherUrl = "https://api.open-meteo.com/v1/forecast?latitude=$lat&longitude=$lon&current=temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m"
                    val weatherResponse = fetchUrl(weatherUrl)

                    // Parse weather response
                    val tempMatch = Regex("\"temperature_2m\":\\s*(-?\\d+\\.?\\d*)").find(weatherResponse)
                    val humidityMatch = Regex("\"relative_humidity_2m\":\\s*(\\d+)").find(weatherResponse)
                    val windMatch = Regex("\"wind_speed_10m\":\\s*(-?\\d+\\.?\\d*)").find(weatherResponse)
                    val codeMatch = Regex("\"weather_code\":\\s*(\\d+)").find(weatherResponse)

                    val temperature = tempMatch?.groupValues?.get(1)?.toDoubleOrNull() ?: 0.0
                    val humidity = humidityMatch?.groupValues?.get(1)?.toIntOrNull() ?: 0
                    val windSpeed = windMatch?.groupValues?.get(1)?.toDoubleOrNull() ?: 0.0
                    val weatherCode = codeMatch?.groupValues?.get(1)?.toIntOrNull() ?: 0

                    val condition = when (weatherCode) {
                        0 -> "Clear sky"
                        1, 2, 3 -> "Partly cloudy"
                        45, 48 -> "Foggy"
                        51, 53, 55 -> "Drizzle"
                        61, 63, 65 -> "Rain"
                        71, 73, 75 -> "Snow"
                        80, 81, 82 -> "Rain showers"
                        95, 96, 99 -> "Thunderstorm"
                        else -> "Unknown"
                    }

                    mapOf(
                        "location" to ToolValue.StringValue(resolvedName),
                        "temperature_celsius" to ToolValue.NumberValue(temperature),
                        "temperature_fahrenheit" to ToolValue.NumberValue(temperature * 9/5 + 32),
                        "humidity_percent" to ToolValue.NumberValue(humidity.toDouble()),
                        "wind_speed_kmh" to ToolValue.NumberValue(windSpeed),
                        "condition" to ToolValue.StringValue(condition)
                    )
                }
            } catch (e: TimeoutCancellationException) {
                Timber.w("Weather API request timed out for location: $location")
                mapOf(
                    "error" to ToolValue.StringValue("Weather API request timed out. Please try again."),
                    "location" to ToolValue.StringValue(location)
                )
            } catch (e: Exception) {
                Timber.e(e, "Weather fetch failed")
                mapOf(
                    "error" to ToolValue.StringValue("Failed to fetch weather: ${e.message}"),
                    "location" to ToolValue.StringValue(location)
                )
            }
        }
    }

    private fun fetchUrl(urlString: String): String {
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = 10000
        connection.readTimeout = 10000

        return try {
            connection.inputStream.bufferedReader().use { it.readText() }
        } finally {
            connection.disconnect()
        }
    }

    /**
     * Evaluate a math expression
     */
    private fun evaluateMathExpression(expression: String): Map<String, ToolValue> {
        return try {
            // Clean the expression
            val cleaned = expression
                .replace("=", "")
                .replace("x", "*")
                .replace("×", "*")
                .replace("÷", "/")
                .trim()

            // Simple expression evaluator (handles basic math)
            val result = evaluateSimpleExpression(cleaned)

            mapOf(
                "result" to ToolValue.NumberValue(result),
                "expression" to ToolValue.StringValue(expression)
            )
        } catch (e: Exception) {
            mapOf(
                "error" to ToolValue.StringValue("Could not evaluate expression: $expression"),
                "expression" to ToolValue.StringValue(expression)
            )
        }
    }

    /**
     * Token parser with index-based iteration supporting peek operations.
     * Enables lookahead for recursive descent parsing without consuming tokens.
     */
    private class TokenParser(private val tokens: List<String>) {
        private var index = 0

        /** Returns true if there are more tokens to consume */
        fun hasNext(): Boolean = index < tokens.size

        /** Returns and consumes the next token */
        fun next(): String {
            if (!hasNext()) throw NoSuchElementException("No more tokens")
            return tokens[index++]
        }

        /** Returns the next token without consuming it, or null if no more tokens */
        fun peek(): String? = if (hasNext()) tokens[index] else null
    }

    /**
     * Simple recursive descent parser for math expressions
     */
    private fun evaluateSimpleExpression(expr: String): Double {
        val tokens = tokenize(expr)
        val parser = TokenParser(tokens)
        return parseExpression(parser)
    }

    private fun tokenize(expr: String): List<String> {
        val tokens = mutableListOf<String>()
        var current = StringBuilder()

        for (char in expr) {
            when {
                char.isDigit() || char == '.' -> current.append(char)
                char in "+-*/()" -> {
                    if (current.isNotEmpty()) {
                        tokens.add(current.toString())
                        current = StringBuilder()
                    }
                    tokens.add(char.toString())
                }
                char.isWhitespace() -> {
                    if (current.isNotEmpty()) {
                        tokens.add(current.toString())
                        current = StringBuilder()
                    }
                }
            }
        }
        if (current.isNotEmpty()) {
            tokens.add(current.toString())
        }
        return tokens
    }

    private fun parseExpression(parser: TokenParser): Double {
        var left = parseTerm(parser)
        while (parser.hasNext()) {
            val op = parser.peek() ?: break
            if (op != "+" && op != "-") break
            parser.next() // consume the operator
            val right = parseTerm(parser)
            left = if (op == "+") left + right else left - right
        }
        return left
    }

    private fun parseTerm(parser: TokenParser): Double {
        var left = parseFactor(parser)
        while (parser.hasNext()) {
            val op = parser.peek() ?: break
            if (op != "*" && op != "/") break
            parser.next() // consume the operator
            val right = parseFactor(parser)
            left = if (op == "*") left * right else left / right
        }
        return left
    }

    private fun parseFactor(parser: TokenParser): Double {
        if (!parser.hasNext()) return 0.0
        val token = parser.next()
        return when {
            token == "(" -> {
                val result = parseExpression(parser)
                if (parser.hasNext()) parser.next() // consume ')'
                result
            }
            token == "-" -> -parseFactor(parser)
            else -> token.toDoubleOrNull() ?: 0.0
        }
    }

    override fun onCleared() {
        super.onCleared()
        synchronized(Companion) {
            if (instance === this) {
                instance = null
            }
        }
    }
}

File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/settings/tool_settings_view_model.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:flutter/foundation.dart';
import 'package:runanywhere/public/runanywhere_tool_calling.dart';
import 'package:runanywhere/public/types/tool_calling_types.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

/// Tool Settings ViewModel (mirroring iOS ToolSettingsViewModel)
///
/// Manages tool calling state and registered tools.
class ToolSettingsViewModel extends ChangeNotifier {
  // Singleton pattern (matches iOS)
  static final ToolSettingsViewModel shared = ToolSettingsViewModel._internal();

  factory ToolSettingsViewModel() => shared;

  ToolSettingsViewModel._internal() {
    unawaited(_loadSettings());
  }

  // State
  List<ToolDefinition> _registeredTools = [];
  bool _toolCallingEnabled = false;

  // Getters
  List<ToolDefinition> get registeredTools => _registeredTools;
  bool get toolCallingEnabled => _toolCallingEnabled;

  set toolCallingEnabled(bool value) {
    _toolCallingEnabled = value;
    unawaited(_saveSettings());
    notifyListeners();
  }

  Future<void> _loadSettings() async {
    final prefs = await SharedPreferences.getInstance();
    _toolCallingEnabled = prefs.getBool('tool_calling_enabled') ?? false;
    await refreshRegisteredTools();
    notifyListeners();
  }

  Future<void> _saveSettings() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool('tool_calling_enabled', _toolCallingEnabled);
  }

  Future<void> refreshRegisteredTools() async {
    _registeredTools = RunAnywhereTools.getRegisteredTools();
    notifyListeners();
  }

  /// Register demo tools (matches iOS implementation)
  Future<void> registerDemoTools() async {
    // 1. Weather Tool - Uses Open-Meteo API (free, no API key required)
    RunAnywhereTools.registerTool(
      const ToolDefinition(
        name: 'get_weather',
        description:
            'Gets the current weather for a given location using Open-Meteo API',
        parameters: [
          ToolParameter(
            name: 'location',
            type: ToolParameterType.string,
            description: "City name (e.g., 'San Francisco', 'London', 'Tokyo')",
          ),
        ],
      ),
      _fetchWeather,
    );

    // 2. Time Tool - Real system time with timezone
    RunAnywhereTools.registerTool(
      const ToolDefinition(
        name: 'get_current_time',
        description: 'Gets the current date, time, and timezone information',
        parameters: [],
      ),
      _getCurrentTime,
    );

    // 3. Calculator Tool - Real math evaluation
    RunAnywhereTools.registerTool(
      const ToolDefinition(
        name: 'calculate',
        description:
            'Performs math calculations. Supports +, -, *, /, and parentheses',
        parameters: [
          ToolParameter(
            name: 'expression',
            type: ToolParameterType.string,
            description: "Math expression (e.g., '2 + 2 * 3', '(10 + 5) / 3')",
          ),
        ],
      ),
      _calculate,
    );

    await refreshRegisteredTools();
  }

  Future<void> clearAllTools() async {
    RunAnywhereTools.clearTools();
    await refreshRegisteredTools();
  }

  // MARK: - Tool Executors

  /// Weather tool executor - fetches real weather data from Open-Meteo
  Future<Map<String, ToolValue>> _fetchWeather(
    Map<String, ToolValue> args,
  ) async {
    final rawLocation = args['location']?.stringValue;
    
    // Require location argument - no hardcoded defaults
    if (rawLocation == null || rawLocation.isEmpty) {
      return {
        'error': const StringToolValue('Missing required argument: location'),
      };
    }
    
    // Clean up location string - Open-Meteo works better with just city names
    // Remove common suffixes like ", CA", ", US", ", USA", etc.
    final location = _cleanLocationString(rawLocation);

    try {
      // Step 1: Geocode the location
      final geocodeUrl = Uri.parse(
        'https://geocoding-api.open-meteo.com/v1/search?name=${Uri.encodeComponent(location)}&count=5&language=en&format=json',
      );

      final geocodeResponse = await http.get(geocodeUrl);
      if (geocodeResponse.statusCode != 200) {
        throw Exception('Geocoding failed');
      }

      final geocodeData =
          jsonDecode(geocodeResponse.body) as Map<String, dynamic>;
      final results = geocodeData['results'] as List?;
      if (results == null || results.isEmpty) {
        return {
          'error': StringToolValue('Could not find location: $location'),
          'location': StringToolValue(location),
        };
      }

      final firstResult = results[0] as Map<String, dynamic>;
      final lat = firstResult['latitude'] as num;
      final lon = firstResult['longitude'] as num;
      final cityName = firstResult['name'] as String? ?? location;

      // Step 2: Fetch weather for coordinates
      final weatherUrl = Uri.parse(
        'https://api.open-meteo.com/v1/forecast?latitude=$lat&longitude=$lon&current=temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m&temperature_unit=fahrenheit&wind_speed_unit=mph',
      );

      final weatherResponse = await http.get(weatherUrl);
      if (weatherResponse.statusCode != 200) {
        throw Exception('Weather fetch failed');
      }

      final weatherData =
          jsonDecode(weatherResponse.body) as Map<String, dynamic>;
      final current = weatherData['current'] as Map<String, dynamic>;
      final temp = current['temperature_2m'] as num? ?? 0;
      final humidity = current['relative_humidity_2m'] as num? ?? 0;
      final windSpeed = current['wind_speed_10m'] as num? ?? 0;
      final weatherCode = current['weather_code'] as int? ?? 0;

      return {
        'location': StringToolValue(cityName),
        'temperature': NumberToolValue(temp.toDouble()),
        'unit': const StringToolValue('fahrenheit'),
        'humidity': NumberToolValue(humidity.toDouble()),
        'wind_speed_mph': NumberToolValue(windSpeed.toDouble()),
        'condition': StringToolValue(_weatherCodeToCondition(weatherCode)),
      };
    } catch (e) {
      return {
        'error': StringToolValue('Weather fetch failed: $e'),
        'location': StringToolValue(location),
      };
    }
  }

  String _weatherCodeToCondition(int code) {
    switch (code) {
      case 0:
        return 'Clear sky';
      case 1:
        return 'Mainly clear';
      case 2:
        return 'Partly cloudy';
      case 3:
        return 'Overcast';
      case 45:
      case 48:
        return 'Foggy';
      case 51:
      case 53:
      case 55:
        return 'Drizzle';
      case 56:
      case 57:
        return 'Freezing drizzle';
      case 61:
      case 63:
      case 65:
        return 'Rain';
      case 66:
      case 67:
        return 'Freezing rain';
      case 71:
      case 73:
      case 75:
        return 'Snow';
      case 77:
        return 'Snow grains';
      case 80:
      case 81:
      case 82:
        return 'Rain showers';
      case 85:
      case 86:
        return 'Snow showers';
      case 95:
        return 'Thunderstorm';
      case 96:
      case 99:
        return 'Thunderstorm with hail';
      default:
        return 'Unknown';
    }
  }

  /// Clean location string for better geocoding results
  /// Removes common suffixes like ", CA", ", US", state abbreviations, etc.
  String _cleanLocationString(String location) {
    var cleaned = location.trim();
    
    // Common patterns to remove: ", CA", ", NY", ", US", ", USA", ", United States"
    // Also handle variations like "CA" at the end
    final patterns = [
      RegExp(r',\s*(US|USA|United States)$', caseSensitive: false),
      RegExp(r',\s*[A-Z]{2}$'), // State abbreviations like ", CA", ", NY"
      RegExp(r',\s*[A-Z]{2},\s*(US|USA)$', caseSensitive: false), // ", CA, US"
    ];
    
    for (final pattern in patterns) {
      cleaned = cleaned.replaceAll(pattern, '');
    }
    
    // Also handle "SF" -> "San Francisco" for common abbreviations
    final abbreviations = {
      'SF': 'San Francisco',
      'NYC': 'New York City',
      'LA': 'Los Angeles',
      'DC': 'Washington DC',
    };
    
    final upperCleaned = cleaned.toUpperCase();
    if (abbreviations.containsKey(upperCleaned)) {
      return abbreviations[upperCleaned]!;
    }
    
    return cleaned;
  }

  /// Time tool executor
  Future<Map<String, ToolValue>> _getCurrentTime(
    Map<String, ToolValue> args,
  ) async {
    final now = DateTime.now();

    return {
      'datetime': StringToolValue(now.toString()),
      'time': StringToolValue(
        '${now.hour.toString().padLeft(2, '0')}:${now.minute.toString().padLeft(2, '0')}:${now.second.toString().padLeft(2, '0')}',
      ),
      'timestamp': StringToolValue(now.toIso8601String()),
      'timezone': StringToolValue(now.timeZoneName),
      'utc_offset': StringToolValue(
        '${now.timeZoneOffset.isNegative ? '-' : '+'}${now.timeZoneOffset.inHours.abs().toString().padLeft(2, '0')}:${(now.timeZoneOffset.inMinutes.abs() % 60).toString().padLeft(2, '0')}',
      ),
    };
  }

  /// Calculator tool executor
  Future<Map<String, ToolValue>> _calculate(
    Map<String, ToolValue> args,
  ) async {
    // Try both 'expression' and 'input' keys - no hardcoded defaults
    final expression = args['expression']?.stringValue ?? args['input']?.stringValue;
    
    if (expression == null || expression.isEmpty) {
      return {
        'error': const StringToolValue('Missing required argument: expression'),
      };
    }

    try {
      // Clean the expression
      final cleanedExpression = expression
          .replaceAll('=', '')
          .replaceAll('x', '*')
          .replaceAll('×', '*')
          .replaceAll('÷', '/')
          .trim();

      final result = _evaluateExpression(cleanedExpression);

      return {
        'result': NumberToolValue(result),
        'expression': StringToolValue(expression),
      };
    } catch (e) {
      return {
        'error': StringToolValue('Could not evaluate expression: $expression'),
        'expression': StringToolValue(expression),
      };
    }
  }

  double _evaluateExpression(String expr) {
    // Remove spaces
    expr = expr.replaceAll(' ', '');

    // Handle power operator first (** or ^)
    if (expr.contains('**')) {
      final parts = expr.split('**');
      var result = double.parse(parts[0]);
      for (var i = 1; i < parts.length; i++) {
        result = _pow(result, double.parse(parts[i]));
      }
      return result;
    } else if (expr.contains('^')) {
      final parts = expr.split('^');
      var result = double.parse(parts[0]);
      for (var i = 1; i < parts.length; i++) {
        result = _pow(result, double.parse(parts[i]));
      }
      return result;
    }

    // Handle simple operations
    if (expr.contains('+')) {
      final parts = expr.split('+');
      return parts.map((p) => double.parse(p)).reduce((a, b) => a + b);
    } else if (expr.contains('-') && !expr.startsWith('-')) {
      final parts = expr.split('-');
      var result = double.parse(parts[0]);
      for (var i = 1; i < parts.length; i++) {
        result -= double.parse(parts[i]);
      }
      return result;
    } else if (expr.contains('*')) {
      final parts = expr.split('*');
      return parts.map((p) => double.parse(p)).reduce((a, b) => a * b);
    } else if (expr.contains('/')) {
      final parts = expr.split('/');
      var result = double.parse(parts[0]);
      for (var i = 1; i < parts.length; i++) {
        result /= double.parse(parts[i]);
      }
      return result;
    }

    return double.parse(expr);
  }
  
  double _pow(double base, double exponent) {
    return math.pow(base, exponent).toDouble();
  }
}
